flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamInputProcessor.java

```diff
@@ -174,6 +179,18 @@ private void initializeNumRecordsIn() {
 		}
 	}
 
+	private boolean checkFinished() throws Exception {
+		boolean isFinished = input.isFinished();
+		if (isFinished) {
+			if (streamOperator instanceof BoundedOneInput) {
```

Comment:
> Why here are you endInput for the head operator and inside the task for the rest of the operators?

The reason is that we don't want `xInputProcessor` to see `OperatorChain`. If we change this point and let `xInputProcessor` see `OperatorChain`, we can do that like you said.

Reply:
If let `xInputProcessor` see non-head operators, we can use the following structure. (`OperatorChain` is still invisible to `XInputProcessor`)

```
============== xInputProcessor.java ==============

private boolean checkFinished() throws Exception {
                        ......
    // invoke endInput(...) of the head operator
    ......
   // invoke endInput(...) of other operators
   streamTask.endNonHeadOperatorInputs()
}

============== xStreamTask.java ==============

public void endNonHeadOperatorInputs throws Exception {
    operatorChain.endNonHeadOperatorInputs();
}

============== OperatorChain.java ==============

public void endNonHeadOperatorInputs throws Exception {
                       ......
}

```

Reply:
It would be better to have this logic in one place and not distributed across two different classes (head is being marked as ended in the processor, while non heads are marked via task).

Also what you proposed adds an (additional) cyclic dependency between Task and Processor, which is also not good.

Maybe the best approach would be to encapsulate this whole logic in `OperatorChain` class and pass it to the processors? For example by adding `OperatorChain#endInput(int inputId)` method, which would:
- if the head is `TwoInput`, it would track which inputs have ended. If all of them, then it would propagate the `endInput` call down the operator chain.
- if the head is `OneInput`, it would just end all of the inputs

For the future, we can not be sure how will be `TwoInputStreamOperator` chaining implemented, but if we keep `OperatorChain` class, the support for propagating `endInput` would need to be adjusted only there.